app/frontend/testnet/testnet-actions.ts

```diff
@@ -0,0 +1,167 @@
+import { initialState } from '../store'
+import {State, Ada, Lovelace} from '../state'
+import {TestnetWallet, getValidStakePools} from './testnet-wallet'
+import WalletLoadingErrorModal from '../components/pages/login/walletLoadingErrorModal'
+// import { getValidStakePools } from './blockchain-testnet-explorer'
+// import BlockchainExplorer from './blockchain-testnet-explorer'
+
+const debounceEvent = (callback, time) => {
+  let interval
+  return (...args) => {
+    clearTimeout(interval)
+    interval = setTimeout(() => {
+      interval = null
+      callback(...args)
+    }, time)
+  }
+}
+
+type SetStateFn = (newState: Partial<State>) => void
+type GetStateFn = () => State
+export default ({setState, getState}: {setState: SetStateFn; getState: GetStateFn}) => {
+
+  const loadWallet = async(state, {cryptoProviderType, walletSecretDef}) => {
+    const wallet = await TestnetWallet(
+      {walletSecretDef}
+    )
+    const accountInfo = await wallet.getAccountStatus(walletSecretDef.rootSecret.pubkeyHex)
+    const delegationHistory = await wallet.getDelegationHistory()
+    const validStakepools = await getValidStakePools()
+    const displayStakingPage = true
+    setState({
+      validStakepools,
+      walletIsLoaded: true,
+      ...accountInfo,
+      delegationHistory,
+      displayStakingPage,
+    })
+  }
+
+  const submitTransaction = (state) => {
+
+  }
+
+  const changeDelegation = async (state) => {
+    console.log('changing delegation')
+  }
+
+  const reloadWalletInfo = () => {
+    return null
+  }
+
+  const calculateDelegatedPercent = (pools) => {
+    return pools.map((pool) => pool.percent).reduce((x, y) => x + y, 0)
+  }
+
+  // DUMMY
+  const fetchPoolInfo = (poolId) => {
+    return poolId === 'adalite-stake-pool-id'
+      && {
+        name: 'AdaLite Stake Pool',
+      }
+  }
+
+  // DUMMY
+  const calculateDelegationFee = () => {
+    const state = getState()
+    setState({
+      delegationFee: state.stakePools.length,
+    })
+    setState({
+      calculatingDelegationFee: false,
+    })
+  }
+
+  const debouncedCalculateDelegationFee = debounceEvent(calculateDelegationFee, 2000)
+
+  // DUMMY
+  const validateDelegationAndCalculateDelegationFee = () => {
+    const state = getState()
+    const isValid =
+      state.stakePools.every((pool) => pool.valid && pool.percent > 0) &&
+      calculateDelegatedPercent(state.stakePools) === 100
+
+    setState({ isDelegationValid: isValid })
+    if (isValid) {
+      setState({ calculatingDelegationFee: true })
+      debouncedCalculateDelegationFee()
+    } else {
+      setState({ delegationFee: 0 })
+    }
+  }
+
+  const updateStakePoolId = (state, e) => {
+    const poolId = e.target.value
+    setState({
+      stakePools: state.stakePools.map((pool, i) => {
+        return Object.assign(
```

Comment:
Why do we need object assign at all if `simple .map(i == x ? orig : new)` would suffice?